packages/modules/fulfillment/src/utils/events.ts

```diff
@@ -0,0 +1,157 @@
+import { FulfillmentSet, GeoZone, ServiceZone } from "@models"
+import { Context, EventBusTypes } from "@medusajs/types"
+import { CommonEvents, FulfillmentUtils, Modules } from "@medusajs/utils"
+
+export function buildCreatedFulfillmentSetEvents({
+  fulfillmentSets,
+  sharedContext,
+}: {
+  fulfillmentSets: FulfillmentSet[]
+  sharedContext: Context
+}): void {
+  if (!fulfillmentSets.length) {
+    return
+  }
+
+  const serviceZones: ServiceZone[] = []
+
+  fulfillmentSets.forEach((fulfillmentSet) => {
+    if (!fulfillmentSet.service_zones?.length) {
+      return
+    }
+
+    serviceZones.push(...fulfillmentSet.service_zones)
+  })
+
+  buildFulfillmentSetEvents({
+    action: CommonEvents.CREATED,
+    fulfillmentSets,
+    sharedContext,
+  })
+
+  buildCreatedServiceZoneEvents({ serviceZones, sharedContext })
+}
+
+export function buildCreatedServiceZoneEvents({
+  serviceZones,
+  sharedContext,
+}: {
+  serviceZones: ServiceZone[]
+  sharedContext: Context
+}): void {
+  if (!serviceZones.length) {
+    return
+  }
+
+  const geoZones: GeoZone[] = []
+
+  serviceZones.forEach((serviceZone) => {
+    if (!serviceZone.geo_zones.length) {
+      return
+    }
+
+    geoZones.push(...serviceZone.geo_zones)
+  })
+
+  buildServiceZoneEvents({
+    action: CommonEvents.CREATED,
+    serviceZones,
+    sharedContext,
+  })
+
+  buildGeoZoneEvents({
+    action: CommonEvents.CREATED,
+    geoZones,
+    sharedContext,
+  })
+}
+
+export function buildFulfillmentSetEvents({
+  action,
+  fulfillmentSets,
+  sharedContext,
+}: {
+  action: string
+  fulfillmentSets: FulfillmentSet[]
+  sharedContext: Context
+}) {
+  if (!fulfillmentSets.length) {
+    return
+  }
+
+  const aggregator = sharedContext.messageAggregator!
+  const messages: EventBusTypes.RawMessageFormat[] = []
+
+  fulfillmentSets.forEach((fulfillmentSet) => {
+    messages.push({
+      service: Modules.FULFILLMENT,
```

Comment:
question: I cannot recall, what purpose does `service` serve?

Reply:
It is actually used no where, it is more of an extra information on the provenance of the event. But we could start with making those metadata optional and re add them later if you prefer. But at least the idea is that you can provide some extra metadata